### PR TITLE
Refactor global.err to test error properties

### PR DESCRIPTION
**Pull request by meeber**
- Add feature https://github.com/chaijs/chai/issues/675
- Fix bug https://github.com/chaijs/chai/issues/648
- Allow object overload of global.err's second argument
- Combine global.err and window.err into one file and function
- Add tests for global.err

### CODE REVIEW

**Comment by meeber** [comment]
One of the tests is passing in node.js but failing in karma due to an added "...".

```js
err(function () {
  err(function () { throw new Err('cat') }, {text: 'cat'});
}, 'expected { Object (message, showDiff) } to have a property \'text\'');
```
```
expected 'expected { Object (message, showDiff, ...) } to have a property \'text\'' to equal 'expected { Object (message, showDiff) } to have a property \'text\''
```

Haven't been able to investigate deeply into it yet. Played around briefly with `truncateThreshold` with no luck.

**Comment by keithamus** [comment]
You could use a regexp to get around that 😄

**Comment by meeber** [comment]
:D I switched that test to a regex and it passes now. Is the engine-specific message inconsistency something we should be concerned about though as a separate issue?

Couple notes about this PR:

- I didn't realize at first that the `err` function had two identical implementations in separate files, one for node and one for karma. I tried a few different approaches for combining them and this felt like the least-awkward of the bunch. Thoughts?

- I originally wanted to use `includes` for the object property testing, but the lack of deep equality was problematic. Some of the assertions create new objects as their `expected` and `actual`, so I think deep equality is the only way to go.

Ended up doing this:
```js
case 'object': return Object.keys(val).forEach(function (key) {
  chai.expect(err).to.have.property(key).and.to.deep.equal(val[key]);
});
```

But I wanted to do this:
```js
case 'object': return chai.expect(err).to.deep.include(val);
```

**Comment by keithamus** [comment]
> Is the engine-specific message inconsistency something we should be concerned about though as a separate issue?

Speaking long term - lots of this plumbing will slowly be refactored out and so there is no need to be concerned about it now.

> I didn't realize at first that the err function had two identical implementations in separate files, one for node and one for karma. I tried a few different approaches for combining them and this felt like the least-awkward of the bunch. Thoughts?

What you've done here is great! Good work!

> I originally wanted to use includes for the object property testing, but the lack of deep equality was problematic. Some of the assertions create new objects as their expected and actual, so I think deep equality is the only way to go.

You can use `.deep.members` I believe. 

```js
case 'object': return chai.expect(err).to.deep.include.members(val);
```

Overall though, this is a great changeset! LGTM - I'd like @lucasfcosta to review also though 😄

File: test/bootstrap/index.js [hunk 1]
```diff
@@ -1,22 +1,39 @@
...
 
-/*!
- * Provide check for fail function.
+/**
+ * Validate that the given function throws an error. Optionally validate some
+ * additional properties of the error:
+ *
+ * If val is a string, validate val equals the error's .message
+ * If val is a regex, validate val matches the error's .message
+ * If val is an object, validate val's props are included in the error object
```
Comment (lucasfcosta) [inline test/bootstrap/index.js]: Great behavior, nice idea

**Comment by lucasfcosta** [comment]
Great work! :smile: 
You've been sending awesome PRs!

LGTM.

**Pull request merged**